### ISSUE TITLE
explicit tail calls via `recur`

### ISSUE THREAD

**Issue by michaelficarra** [jashkenas/coffeescript#2665]
I've been meaning to open this issue for months. And now that ES6 has proper tail calls, allowing TCO in implementations, it's more appropriate than ever.

Clojure has [this really neat syntactic feature, `recur`](http://clojure.org/special_forms#Special%20Forms--%28recur%20exprs*%29), that performs a recursive call, but is only valid in tail position. That's really useful if you make changes later, accidentally moving the recursion out of tail position, preventing the implementation from performing TCO. Your CoffeeScript compiler should fail to compile your program rather than letting you run into stack overflows at runtime (assuming you hit that scenario before your users do!). This would be a great feature for CoffeeScript, no?

Possible downsides:
- Backwards incompatible with programs that use `recur` as an identifier.
- Must generate a reference to the containing function
- Nothing else!

Sample compilation:

``` coffee
fact = do ->
  tailFact = (n, a) ->
    return a if n is 0
    recur (n - 1), (n * a)
  (x) -> tailFact x, 1
```

``` js
fact = (function() {
  var tailFact;
  tailFact = function recur$0(n, a) {
    if (n === 0) {
      return a;
    }
    return recur$0(n - 1, n * a);
  };
  return function(x) {
    return tailFact(x, 1);
  };
})();
```

**tl;dr** `recur` should be a reference to the containing function, and the compiler should enforce that it is in tail position.

**edit:** Fixed example.
